### ISSUE TITLE
go/build: Quadratic complexity in Context.Import

### ISSUE THREAD

**Issue by catenacyber** [golang/go#63966]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.21 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

It happens only on gotip, not on go 1.21

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/root/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/root/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/root/.go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/root/.go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.21"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/src/ngolo-fuzzing/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2481516251=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Run https://go.dev/play/p/R_pJ6FL70Ok

### What did you expect to see?

The program finishing and printing Hello

### What did you see instead?

`timeout running program`

Found by https://github.com/catenacyber/ngolo-fuzzing with oss-fuzz : 
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=63873

This looks due to a quadratic complexity between the number of directories in GOPATH for the build.Context (which do not seem to get deduplicated) and the number of colon-separated entries in srcDir argument for https://pkg.go.dev/go/build@go1.21.3#Context.Import
The documentation does not seem to mention that `srcDir` gets interpreted as a colon-separated list of entries...
Is this a problem ?

**Comment by bcmills** [comment]
> Found by https://github.com/catenacyber/ngolo-fuzzing with oss-fuzz :
> https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=63873

I appear not to have the permissions needed to view that issue. Can you summarize it here?

**Comment by catenacyber** [comment]
> I appear not to have the permissions needed to view that issue. Can you summarize it here?

There is nothing more there beyond the reproducer https://go.dev/play/p/R_pJ6FL70Ok

> How would the GOPATH field of the build.Context end up set to a long list of invalid or duplicated paths in the first place? (What is the threat model that results in treating quadratic complexity here as a problem?)

I am not sure there is an issue indeed... But it looks a bit fishy (I do not see any warning in the documentation, and this is the first time ngolo-fuzzing finds something for `go/build`)

> I don't see how it would end up being treated as such a list — what leads you to believe that it does?

If the value gets shorter, we do not get the timeout any longer...
If GOPATH gets shorter, we do not get the timeout any longer...
So, it looks to me that there is a multiplying effect between both... but I did not check the code...

**Comment by bcmills** [comment]
> So, it looks to me that there is a multiplying effect between both... but I did not check the code...

I would guess that the quadratic behavior comes from the calls to `filepath.EvalSymlinks` in `hasSubdir`, although of course I would want to see a profile or trace to confirm that guess.
(In my experience, calls to `EvalSymlinks` are usually not the best solution to problems, but in this case they've been there for a long time — I don't think we can just rip them out without a proposal and some analysis.)

Since we don't expect `go/build` to be used with an attacker-controlled `GOPATH`, I don't plan to make any changes here. We could probably accept a fix if someone wants to gather a profile or set up a benchmark and can identify a fix that isn't too invasive, but this probably isn't worth adding much (if any!) code complexity to address.

**Comment by catenacyber** [comment]
I ran a profile.
Most time is spent in 
`go/build.(*Context).Import` doing a first loop over gopath 
```
for _, root := range gopath { 
    if searchVendor(root, false) { 
```

Then searchVendor looping over the srcDir being split over `/
```
for { 
    vendor := ctxt.joinPath(root, sub, "vendor") 
    if ctxt.isDir(vendor) { 
...
    i := strings.LastIndex(sub, "/")
    if i < 0 {
        break
    }					
    sub = sub[:i]
```

I am ok for closing this
